src/plugins/vis_type_xy/public/editor/components/options/metrics_axes/index.tsx

```diff
@@ -170,33 +171,58 @@ function MetricsAxisOptions(props: ValidationVisOptionsProps<BasicVislibParams>)
   );
 
   const onValueAxisPositionChanged = useCallback(
-    (index: number, value: ValueAxis['position']) => {
+    (index: number, axisPosition: ValueAxis['position']) => {
+      const isHorizontalAxis = isAxisHorizontal(axisPosition);
       const valueAxes = [...stateParams.valueAxes];
-      const name = getUpdatedAxisName(value, valueAxes);
+      const name = getUpdatedAxisName(axisPosition, valueAxes);
+      const [categoryAxes] = stateParams.categoryAxes;
 
-      valueAxes[index] = {
-        ...valueAxes[index],
-        name,
-        position: value,
-      };
-      setValue('valueAxes', valueAxes);
+      if (isAxisHorizontal(categoryAxes.position) === isHorizontalAxis) {
+        const updatedCategoryAxes = {
+          ...categoryAxes,
+          position: mapPosition(categoryAxes.position),
+        };
+
+        setValue('categoryAxes', [updatedCategoryAxes]);
+
+        const oldPosition = valueAxes[index].position;
+        const newValueAxes = valueAxes.map(({ position, ...axis }, i) => ({
+          ...axis,
+          position:
+            i === index
+              ? axisPosition
+              : mapPositionOpposingOpposite(position, oldPosition, axisPosition),
+        }));
+        setValue('valueAxes', newValueAxes);
+      } else {
+        valueAxes[index] = {
+          ...valueAxes[index],
+          name,
+          position: axisPosition,
+        };
+        setValue('valueAxes', valueAxes);
+      }
     },
-    [stateParams.valueAxes, setValue]
+    [stateParams.valueAxes, stateParams.categoryAxes, setValue]
   );
 
   const onCategoryAxisPositionChanged = useCallback(
-    (chartPosition: Axis['position']) => {
-      const isChartHorizontal = isAxisHorizontal(chartPosition);
-      setIsCategoryAxisHorizontal(isAxisHorizontal(chartPosition));
-
-      stateParams.valueAxes.forEach((axis, index) => {
-        if (isAxisHorizontal(axis.position) === isChartHorizontal) {
-          const position = mapPosition(axis.position);
-          onValueAxisPositionChanged(index, position);
-        }
-      });
+    (axisPosition: CategoryAxis['position']) => {
+      const isHorizontalAxis = isAxisHorizontal(axisPosition);
+
+      if (
+        stateParams.valueAxes.some(
+          ({ position }) => isAxisHorizontal(position) === isHorizontalAxis
+        )
+      ) {
+        const newValueAxes = stateParams.valueAxes.map(({ position, ...axis }) => ({
+          ...axis,
+          position: mapPosition(position),
+        }));
+        setValue('valueAxes', newValueAxes);
+      }
     },
-    [stateParams.valueAxes, onValueAxisPositionChanged]
+    [setValue, stateParams.valueAxes]
```

Comment:
It requires functions too in case the reference to the function changes.

![image](https://user-images.githubusercontent.com/19007109/98274940-ab372380-1f59-11eb-92b7-fddcae4662c1.png)

Reply:
mm, ok